yoyodyne/data/indexes.py

```diff
@@ -4,36 +4,79 @@
 import pickle
 from typing import Dict, List, Optional, Set
 
-from .. import special
+from .. import defaults, special
 
 
-class SymbolMap:
-    """Tracks mapping from index to symbol and symbol to index."""
+class Error(Exception):
+    pass
+
+
+class Index:
+    """Maintains the index over the vocabularies.
+
+    For consistency, one is recommended to lexicographically sort the
+    vocabularies ahead of time."""
 
     index2symbol: List[str]
     symbol2index: Dict[str, int]
 
-    def __init__(self, vocabulary: List[str]):
+    def __init__(
+        self,
+        *,
+        source_vocabulary: List[str],
+        features_vocabulary: Optional[List[str]] = None,
+        target_vocabulary: Optional[List[str]] = None,
+        tie_embeddings: bool = defaults.TIE_EMBEDDINGS,
+    ):
+        """Initializes the index.
+
+        Args:
+            source_vocabulary (List[str]).
+            features_vocabulary (List[str], optional).
+            target_vocabulary (List[str], optional).
+            tie_embeddings: (bool).
+        """
+        super().__init__()
+        self.tie_embeddings = tie_embeddings
+        # We store all separate vocabularies for logging purposes.
```

Comment:
this doesn't look like it's used anywhere

Reply:
This is meant to explain the properties `source_vocabulary, self.target_vocabulary, self.features_vocabulary`, which would not be needed since we use one embedding matrix, but which we store for logging purposes.